### PR TITLE
Multi-Output recipes, and reliable secondary outputs

### PR DESCRIPTION
**Pull request by Mekle001**
# Changes proposed in this pull request:
- Add support for multi-output recipes, so that one recipe can create all variants of one timber block
- Add support for custom recipes to require that the recipe 'exists' to be valid
  - This enables a multi-output recipe to replace a classic recipe in which the inputs match
  - When a multi-output replaces, it also looks for classic recipes who have outputs that are alternate in the multi-recipe and removes them
- Add support for multi-recipes to cycle through the outputs in the recipe list in buildings
- Add ability to capture actual secondary outputs (stay in grid) from recipes as they are being taught, This increases compatibility with many mod recipes, including aquaculture fish fillets.
- Some minor tweaks to display of the request tree in buildings, to make crafting jobs a bit clearer
- Fix Combat Academy and Archery to mark beds as unoccupied on wake up, letting the residents sleep the next night
- use Localizable strings for Combat Academy and Archery
- Ensure that multi recipes replace simple recipes, and that they update when they change
- Change all of the lumberjack recipes to be multi-recipes. They are now Log -> [Stripped Log, Wood, Stripped Wood]

Review please

### CODE REVIEW

File: src/api/java/com/minecolonies/api/inventory/container/ContainerCrafting.java [hunk 1]
```diff
@@ -46,6 +48,11 @@
      */
     private final Slot craftResultSlot;
 
+    /**
+     * The secondary outputs
+     */
+    private final java.util.List<ItemStack> remainingItems;
```
Comment (Raycoms) [inline src/api/java/com/minecolonies/api/inventory/container/ContainerCrafting.java]: java.util.List?

File: src/api/java/com/minecolonies/api/inventory/container/ContainerCrafting.java [hunk 2]
```diff
@@ -365,4 +375,38 @@ public BlockPos getPos()
...
+     * Get for the remaining items. 
+     * @return
+     */
+    public java.util.List<ItemStack> getRemainingItems()
+    {
+        final Optional<ICraftingRecipe> iRecipe = this.world.getRecipeManager().getRecipe(IRecipeType.CRAFTING, craftMatrix, world);
+        RecipeBook r;
+        if(world.isRemote)
+        {
+            r = new ClientRecipeBook(this.world.getRecipeManager());
```
Comment (Raycoms) [inline src/api/java/com/minecolonies/api/inventory/container/ContainerCrafting.java]: this class might not load because it has the ClientRecipeBook in SMP. have you tested this in smp?

Comment (Raycoms) [reply src/api/java/com/minecolonies/api/inventory/container/ContainerCrafting.java]: what do we need it for? Just do check if its unlocked? I don't think we should care about that

File: src/main/java/com/minecolonies/coremod/client/gui/containers/WindowFurnaceCrafting.java [hunk 3]
```diff
@@ -100,7 +101,7 @@ public void onPress(final Button button)
 
                 if (!ItemStackUtils.isEmpty(primaryOutput))
                 {
-                    Network.getNetwork().sendToServer(new AddRemoveRecipeMessage(building, input, 1, primaryOutput, false));
+                    Network.getNetwork().sendToServer(new AddRemoveRecipeMessage(building, input, 1, primaryOutput, null, false));
                 }
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/client/gui/containers/WindowFurnaceCrafting.java:105]: why not an empty list?

File: src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java [hunk 4]
```diff
@@ -197,7 +197,7 @@ public IRecipeStorage getFirstRecipe(final Predicate<ItemStack> stackPredicate)
         for (final IToken<?> token : recipes)
         {
             final IRecipeStorage storage = IColonyManager.getInstance().getRecipeManager().getRecipes().get(token);
-            if (storage != null && stackPredicate.test(storage.getPrimaryOutput()))
+            if (storage != null && (stackPredicate.test(storage.getPrimaryOutput()) || storage.getAlternateOutputs().stream().anyMatch(i -> stackPredicate.test(i))))
             {
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java:201]: Do we wan to check for secondary outputs too?

File: src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java [hunk 5]
```diff
@@ -215,7 +215,12 @@ public IRecipeStorage getFirstRecipe(final Predicate<ItemStack> stackPredicate)
                 final ItemStorage checkItem = foo.getKey().getCleanedInput().stream().max(Comparator.comparingInt(ItemStorage::getAmount)).get();
                 candidates.put(foo.getKey(), getWarehouseCount(checkItem));
             }
-            return candidates.entrySet().stream().min(Map.Entry.comparingByValue(Comparator.reverseOrder())).get().getKey();
+            firstFound = candidates.entrySet().stream().min(Map.Entry.comparingByValue(Comparator.reverseOrder())).get().getKey();
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java]: Either we should "break;" at first found, or do something else, because we override firstfound with each iteration.

File: src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java [hunk 6]
```diff
@@ -851,16 +857,34 @@ public void checkForWorkerSpecificRecipes()
...
                         }
                     }
                 }
-                if(!duplicateFound)
+                if(duplicateFound == null)
                 {
                     addRecipeToList(recipeToken);    
                     colony.getRequestManager().onColonyUpdate(request -> request.getRequest() instanceof IDeliverable && ((IDeliverable) request.getRequest()).matches(recipeStorage.getPrimaryOutput()));
                 }
+                else if(newRecipe.isMustExist() && duplicateFound != recipeToken)
```
Comment (Raycoms) [inline src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java]: isMustExist seems weird

File: src/api/java/com/minecolonies/api/inventory/container/ContainerCrafting.java (hunk 1 above)
Comment (Mekle001) [reply src/api/java/com/minecolonies/api/inventory/container/ContainerCrafting.java]: Yes, just List resolves to something else here.

File: src/main/java/com/minecolonies/coremod/client/gui/containers/WindowFurnaceCrafting.java (hunk 3 above)
Comment (Mekle001) [reply src/main/java/com/minecolonies/coremod/client/gui/containers/WindowFurnaceCrafting.java:105]: We could, but it will become an empty list either way due to the recipestorage constructor.

File: src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java (hunk 6 above)
Comment (Mekle001) [reply src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java]: The name? or the functionality?

Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java]: the name

File: src/api/java/com/minecolonies/api/inventory/container/ContainerCrafting.java (hunk 1 above)
Comment (Mekle001) [reply src/api/java/com/minecolonies/api/inventory/container/ContainerCrafting.java]: Ok, I know what happened, and I have a fix.

File: src/api/java/com/minecolonies/api/inventory/container/ContainerCrafting.java (hunk 2 above)
Comment (Mekle001) [reply src/api/java/com/minecolonies/api/inventory/container/ContainerCrafting.java]: That's all it's used for. We probably don't need most of the checks there.

File: src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java (hunk 5 above)
Comment (Mekle001) [reply src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java]: In the scan loop, we check against null before setting it, This highlighted one is in an "If", after populating counts, so we're only overriding it once in case we're adjusting for warehouse inventory. 
This is followed by another if, where we check for a multi-output, and convert it to a simple if so. 

That's a total of 3 sets maximum.

File: src/api/java/com/minecolonies/api/inventory/container/ContainerCrafting.java (hunk 2 above)
Comment (Raycoms) [reply src/api/java/com/minecolonies/api/inventory/container/ContainerCrafting.java]: yeah, remove it then. too error prone having this here.

File: src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java (hunk 5 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java]: But then "firstFound" is not exactly the right name, right?

File: src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java (hunk 6 above)
Comment (Mekle001) [reply src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java]: Got it. Happy to 'break the standard' here of 'is' for boolean and switch to 'get'

File: src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java (hunk 5 above)
Comment (Mekle001) [reply src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java]: Fair point

File: src/api/java/com/minecolonies/api/inventory/container/ContainerCrafting.java (hunk 2 above)
Comment (Mekle001) [reply src/api/java/com/minecolonies/api/inventory/container/ContainerCrafting.java]: Gone

File: src/main/java/com/minecolonies/coremod/client/gui/containers/WindowFurnaceCrafting.java (hunk 3 above)
Comment (Mekle001) [reply src/main/java/com/minecolonies/coremod/client/gui/containers/WindowFurnaceCrafting.java:105]: Went ahead and did an empty list.

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/core/AbstractCraftingProductionResolver.java [hunk 7]
```diff
@@ -148,13 +148,13 @@ protected boolean canBuildingCraftStack(@NotNull final IRequestManager manager,
                 {
                     materialRequests.add(createNewRequestForStack(manager, craftingHelperStack, ingredient.getAmount(), ingredient.getAmount()));
                 }
-                else if (!ItemStackUtils.isEmpty(container) && ItemStackUtils.compareItemStacksIgnoreStackSize(container, craftingHelperStack, false, true))
+                else if (ItemStackUtils.isEmpty(container) || ItemStackUtils.compareItemStacksIgnoreStackSize(container, craftingHelperStack, false, true))
```
Comment (marchermans) [inline src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/core/AbstractCraftingProductionResolver.java]: Why this change? Documentation is missing here. The original code makes more sense....

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAICrafting.java [hunk 8]
```diff
@@ -173,13 +175,13 @@ protected IAIState getRecipe()
             {
                 remaining = inputStorage.getAmount();
             }
-            else if (!ItemStackUtils.isEmpty(container) && ItemStackUtils.compareItemStacksIgnoreStackSize(inputStorage.getItemStack(), container , false, true))
+            else if (ItemStackUtils.isEmpty(container) || !ItemStackUtils.compareItemStacksIgnoreStackSize(inputStorage.getItemStack(), container , false, true))
```
Comment (marchermans) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAICrafting.java]: Again this change. These kind of suttle changes need to be documented if it is not a hotfix.
Why? This logic does not seem to make sense to me.

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAICrafting.java [hunk 9]
```diff
@@ -251,13 +253,13 @@ protected IAIState checkForItems(@NotNull final IRecipeStorage storage)
             {
                 remaining = inputStorage.getAmount();
             }
-            else if(!ItemStackUtils.isEmpty(container) && ItemStackUtils.compareItemStacksIgnoreStackSize(inputStorage.getItemStack(), container , false, true))
+            else if(ItemStackUtils.isEmpty(container) || !ItemStackUtils.compareItemStacksIgnoreStackSize(inputStorage.getItemStack(), container , false, true))
```
Comment (marchermans) [inline src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAICrafting.java]: and again! Here it is even worse, Since here the second part of the OR-Clause has a negation but in the previous cases where you patches this it did not. Why was this changed?

File: src/api/java/com/minecolonies/api/crafting/IRecipeStorage.java [hunk 10]
```diff
@@ -8,12 +8,22 @@
...
 
 /**
  * Interface which describes the RecipeStorage.
  */
 public interface IRecipeStorage
 {
+    /**
+     * Enum of the different types of storage
+     */
+    public enum RecipeStorageType
```
Comment (marchermans) [inline src/api/java/com/minecolonies/api/crafting/IRecipeStorage.java]: Not expandable!
This is part of the API. ENUMS are basically a nogo here.

File: src/api/java/com/minecolonies/api/crafting/IRecipeStorageFactory.java [hunk 11]
```diff
@@ -70,6 +83,11 @@ RecipeStorage getNewInstance(
       @NotNull final List<ItemStack> input,
       final int gridSize,
       @NotNull final ItemStack primaryOutput,
-      @Nullable final Block intermediate);
+      @Nullable final Block intermediate,
+      @Nullable final String source,
```
Comment (marchermans) [inline src/api/java/com/minecolonies/api/crafting/IRecipeStorageFactory.java]: Documentation is not updated. What is source?

File: src/api/java/com/minecolonies/api/crafting/IRecipeStorage.java [hunk 12]
```diff
@@ -72,6 +82,45 @@ default boolean fullFillRecipe(@NotNull final IItemHandler... inventories)
...
+     * @param stackPredicate Predicate to select the right stack
+     * @return the RecipeStorage that is "right" for that output
+     */
+    public RecipeStorage getClassicForMultiOutput(final Predicate<ItemStack> stackPredicate);
+
+    /**
+     * Source of the recipe, ie registry name.
+     * @return
+     */
+    public String getRecipeSource();
```
Comment (marchermans) [inline src/api/java/com/minecolonies/api/crafting/IRecipeStorage.java]: If this is supposed to be a registry name, or reference one, Make it return a ResourceLocation instead,
If the return value is optional, mark it with `@Nullable` or return an optional and make the use of it like a functional invocation.

File: src/api/java/com/minecolonies/api/crafting/RecipeStorage.java [hunk 13]
```diff
@@ -2,24 +2,40 @@
...
 import java.util.List;
+import java.util.function.Predicate;
+import java.util.stream.Collectors;
 
 /**
  * Class used to represent a recipe in minecolonies.
  */
 public class RecipeStorage implements IRecipeStorage
 {
+
```
Comment (marchermans) [inline src/api/java/com/minecolonies/api/crafting/RecipeStorage.java:32]: See comments on the interface for this.

File: src/api/java/com/minecolonies/api/crafting/RecipeStorage.java [hunk 14]
```diff
@@ -216,6 +312,22 @@ public int hashCode()
         result = 31 * result + primaryOutput.hashCode();
         result = 31 * result + (intermediate != null ? intermediate.hashCode() : 0);
         result = 31 * result + gridSize;
+        if(recipeSource != null && !recipeSource.isEmpty())
```
Comment (marchermans) [inline src/api/java/com/minecolonies/api/crafting/RecipeStorage.java]: I am not sure, but this does not look like it was autogenerated by IDEA!

File: src/api/java/com/minecolonies/api/crafting/RecipeStorage.java [hunk 15]
```diff
@@ -228,13 +340,20 @@ public int hashCode()
...
-            final ItemStack container = stack.getItem().getContainerItem(stack);
-            if (!ItemStackUtils.isEmpty(container))
+            for (final ItemStack stack : input)
             {
-                container.setCount(stack.getCount());
-                secondaryStacks.add(container);
+                final ItemStack container = stack.getItem().getContainerItem(stack);
+                if (!ItemStackUtils.isEmpty(container))
+                {
+                    container.setCount(stack.getCount());
```
Comment (marchermans) [inline src/api/java/com/minecolonies/api/crafting/RecipeStorage.java:368]: This does not work........
The API clearly states that the container Item is returned when the stack is consumed! You are not allowed to modify this stack willy nilly since that violates that contract. There might be mods out there which consume an entire stack and only return 1 container item!

File: src/api/java/com/minecolonies/api/crafting/RecipeStorage.java [hunk 16]
```diff
@@ -329,29 +448,36 @@ public boolean fullfillRecipe(final List<IItemHandler> handlers)
...
 
-            final ItemStack container = stack.getItem().getContainerItem(stack);
-            if (!ItemStackUtils.isEmpty(container))
-            {
-                container.setCount(stack.getCount());
-                secondaryStacks.add(container);
+                final ItemStack container = stack.getItem().getContainerItem(stack);
+                if (!ItemStackUtils.isEmpty(container))
+                {
+                    container.setCount(stack.getCount());
```
Comment (marchermans) [inline src/api/java/com/minecolonies/api/crafting/RecipeStorage.java:493]: Again see above this is not a legal use of the API as far as I am concerned.

File: src/api/java/com/minecolonies/api/crafting/RecipeStorage.java [hunk 17]
```diff
@@ -329,29 +448,36 @@ public boolean fullfillRecipe(final List<IItemHandler> handlers)
...
-            if (stack.getItem() == ModItems.buildTool)
+            secondaryStacks.addAll(secondaryOutputs);
+        }
+        else
+        {
+            for (final ItemStack stack : input)
             {
-                continue;
-            }
+                if (stack.getItem() == ModItems.buildTool)
```
Comment (marchermans) [inline src/api/java/com/minecolonies/api/crafting/RecipeStorage.java:485]: You are in the API, no special casing of Minecolonies specific stuff.
Create a new Interface if you want to and smag that on the buildtool then do an instance of.
Or create a behaviour registry.

Regardless either refactor this class so that it is not part of the API or refactor it so that it does not do any special casing.

File: src/api/java/com/minecolonies/api/crafting/RecipeStorage.java [hunk 18]
```diff
@@ -365,4 +491,65 @@ private void insertCraftedItems(final List<IItemHandler> handlers)
             }
         }
     }
+
+    @Override
+    public RecipeStorage getClassicForMultiOutput(final ItemStack requiredOutput)
+    {
+        return StandardFactoryController.getInstance().getNewInstance(
```
Comment (marchermans) [inline src/api/java/com/minecolonies/api/crafting/RecipeStorage.java]: You are in the concrete implementation, you do not need to use the factory controller......
You can invoke yourself just fine,

File: src/api/java/com/minecolonies/api/inventory/container/ContainerCrafting.java [hunk 19]
```diff
@@ -2,6 +2,7 @@
 
 import com.ldtteam.structurize.api.util.ItemStackUtils;
 import com.minecolonies.api.inventory.ModContainers;
+import net.minecraft.client.util.ClientRecipeBook;
```
Comment (marchermans) [inline src/api/java/com/minecolonies/api/inventory/container/ContainerCrafting.java]: Import of Client side only stuff.

File: src/main/java/com/minecolonies/coremod/client/gui/WindowListRecipes.java [hunk 20]
```diff
@@ -111,7 +119,23 @@ public void updateElement(final int index, @NotNull final Pane rowPane)
             {
                 @NotNull final IRecipeStorage recipe = recipes.get(index);
                 final ItemIcon icon = rowPane.findPaneOfTypeByID(OUTPUT_ICON, ItemIcon.class);
-                icon.setItem(recipe.getPrimaryOutput());
+                if(recipe.getRecipeType() == RecipeStorageType.MULTI_OUTPUT)
```
Comment (marchermans) [inline src/main/java/com/minecolonies/coremod/client/gui/WindowListRecipes.java]: This comes with the teritory of writing Enums in the API.
Refactor this so that the recipe type can handle it itself.

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingArchery.java [hunk 21]
```diff
@@ -284,6 +284,30 @@ public BuildingEntry getBuildingRegistryEntry()
...
+            BlockState state = world.getBlockState(pos);
+            state = state.getBlock().getExtendedState(state, world, pos);
+            if (state.getBlock() instanceof BedBlock
+                  && state.get(BedBlock.OCCUPIED)
+                  && state.get(BedBlock.PART).equals(BedPart.HEAD))
+            {
+                world.setBlockState(pos, state.with(BedBlock.OCCUPIED, false), 0x03);
+            }
+        }
+    }
```
Comment (marchermans) [inline src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingArchery.java:309]: Is this not supposed to be default behaviour?
Why was this added here?

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingCombatAcademy.java [hunk 22]
```diff
@@ -378,6 +378,30 @@ public BuildingEntry getBuildingRegistryEntry()
...
+            BlockState state = world.getBlockState(pos);
+            state = state.getBlock().getExtendedState(state, world, pos);
+            if (state.getBlock() instanceof BedBlock
+                  && state.get(BedBlock.OCCUPIED)
+                  && state.get(BedBlock.PART).equals(BedPart.HEAD))
+            {
+                world.setBlockState(pos, state.with(BedBlock.OCCUPIED, false), 0x03);
+            }
+        }
+    }
```
Comment (marchermans) [inline src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingCombatAcademy.java:403]: Again this is supposed to be default behaviour for the entity.
Why was this not handled there.

File: src/main/java/com/minecolonies/coremod/colony/crafting/CustomRecipe.java [hunk 23]
```diff
@@ -234,6 +268,27 @@ public static CustomRecipe parse(@NotNull final JsonObject recipeJson)
             recipe.result = idToItemStack(recipeJson.get(RECIPE_RESULT_PROP).getAsString());
         }
 
+        if (recipeJson.has(RECIPE_ALTERNATE_PROP))
+        {
+            for (JsonElement e : recipeJson.get(RECIPE_ALTERNATE_PROP).getAsJsonArray())
+            {
+                if (e instanceof JsonElement && e.isJsonObject())
```
Comment (marchermans) [inline src/main/java/com/minecolonies/coremod/colony/crafting/CustomRecipe.java]: `e instanceof JsonElement` is always true.

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/requests/StandardRequests.java [hunk 24]
```diff
@@ -343,11 +343,9 @@ protected AbstractCraftingRequest(
         public final ITextComponent getShortDisplayString()
         {
             final ITextComponent result = new NonSiblingFormattingTextComponent();
-            final ITextComponent preType = new TranslationTextComponent(getTranslationKey());
 
-            result.appendSibling(preType);
-
-            preType.appendSibling(getRequest().getStack().getTextComponent());
+            result.appendSibling(new StringTextComponent(getRequest().getMinCount() + " * Recipe:"));
+            result.appendSibling(getRequest().getStack().getTextComponent());
```
Comment (marchermans) [inline src/main/java/com/minecolonies/coremod/colony/requestsystem/requests/StandardRequests.java]: This makes no sense.
What are you trying to achieve here?

Comment (marchermans) [reply src/main/java/com/minecolonies/coremod/colony/requestsystem/requests/StandardRequests.java]: Also not localizable!

**Review by marchermans — CHANGES_REQUESTED** [review]
A lot of stuff in the api is not how it is supposed to be.
A bunch of stuff introduces regressions.
And a LOT LOT LOT of documentation is missing, especially on your suddle changes to how itemstacks and containers are handled.

This needs to be carefully considered.

File: src/api/java/com/minecolonies/api/crafting/IRecipeStorage.java (hunk 12 above)
Comment (Mekle001) [reply src/api/java/com/minecolonies/api/crafting/IRecipeStorage.java]: I thought about that, I'd intended to support a variety of sources, including possibly player name. None of that is done yet though.

File: src/api/java/com/minecolonies/api/crafting/RecipeStorage.java (hunk 15 above)
Comment (Mekle001) [reply src/api/java/com/minecolonies/api/crafting/RecipeStorage.java:368]: Either way this does not work, since we operate off of cleaned inputs. In the case of cake, the stack is 2 milk buckets, and getContainerItem returns one bucket, but vanilla crafting returns 2. The good news is that this code is legacy/compat. With the secondary items list, we return what the recipe says gets returned.

File: src/api/java/com/minecolonies/api/inventory/container/ContainerCrafting.java (hunk 19 above)
Comment (Mekle001) [reply src/api/java/com/minecolonies/api/inventory/container/ContainerCrafting.java]: Missed cleanup from earlier versions.

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingArchery.java (hunk 21 above)
Comment (Mekle001) [reply src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingArchery.java:309]: All of our buildings where citizens sleep have this code. I missed porting it when I enabled sleeping in the Archery and the Combat Academy, which meant that the citizens would sleep one night, and go about there business the next day with the beds still marked occupied. The next night, they couldn't find a bed, so would stand around the hut sleeping.

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingCombatAcademy.java (hunk 22 above)
Comment (Mekle001) [reply src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingCombatAcademy.java:403]: The entities do not handle this for any of our buildings.

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingArchery.java (hunk 21 above)
Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingArchery.java:309]: This will be solved in the future when we modularize this

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/core/AbstractCraftingProductionResolver.java (hunk 7 above)
Comment (Mekle001) [reply src/main/java/com/minecolonies/coremod/colony/requestsystem/resolvers/core/AbstractCraftingProductionResolver.java]: reverted

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAICrafting.java (hunk 8 above)
Comment (Mekle001) [reply src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAICrafting.java]: reverted

File: src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAICrafting.java (hunk 9 above)
Comment (Mekle001) [reply src/main/java/com/minecolonies/coremod/entity/ai/basic/AbstractEntityAICrafting.java]: reverted

File: src/api/java/com/minecolonies/api/crafting/RecipeStorage.java (hunk 14 above)
Comment (Mekle001) [reply src/api/java/com/minecolonies/api/crafting/RecipeStorage.java]: I don't use IDEA, so I was trying to keep in line.

File: src/api/java/com/minecolonies/api/crafting/IRecipeStorageFactory.java (hunk 11 above)
Comment (Mekle001) [reply src/api/java/com/minecolonies/api/crafting/IRecipeStorageFactory.java]: documentation updated

File: src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java (hunk 5 above)
Comment (Mekle001) [reply src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java]: I should note that this has been renamed.

File: src/api/java/com/minecolonies/api/crafting/IRecipeStorage.java (hunk 12 above)
Comment (marchermans) [reply src/api/java/com/minecolonies/api/crafting/IRecipeStorage.java]: Then document it as such.

File: src/api/java/com/minecolonies/api/crafting/RecipeStorage.java (hunk 15 above)
Comment (marchermans) [reply src/api/java/com/minecolonies/api/crafting/RecipeStorage.java:368]: If this is legacy code. Why was it added in your PR?

File: src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingArchery.java (hunk 21 above)
Comment (marchermans) [reply src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingArchery.java:309]: Okey....
Then create an interface with a default set of methods that handle this? ANd smag that on all buildings?+

I see no reason this can not be cleaned up.

Comment (Raycoms) [reply src/main/java/com/minecolonies/coremod/colony/buildings/workerbuildings/BuildingArchery.java:309]: I got a PR going atm that will clean that up. Might need another week to be mature though

File: src/api/java/com/minecolonies/api/crafting/IRecipeStorage.java (hunk 12 above)
Comment (Mekle001) [reply src/api/java/com/minecolonies/api/crafting/IRecipeStorage.java]: I'll go ahead and convert to ResourceLocation, since there are difficulties in overloading it the way I planned with the way I ended up using it.

File: src/api/java/com/minecolonies/api/crafting/RecipeStorage.java (hunk 15 above)
Comment (Mekle001) [reply src/api/java/com/minecolonies/api/crafting/RecipeStorage.java:368]: The container case becomes legacy code with this PR, there was a bug reported around producing too few buckets while I was working on it, so I included it to avoid more merge conflicts.

File: src/api/java/com/minecolonies/api/crafting/RecipeStorage.java (hunk 17 above)
Comment (Mekle001) [reply src/api/java/com/minecolonies/api/crafting/RecipeStorage.java:485]: Ok, the buildTool exception is pre-existing. We make this exception in 3 places in this file.  Changing this item is a Structurize change.

File: src/api/java/com/minecolonies/api/crafting/RecipeStorage.java (hunk 18 above)
Comment (Mekle001) [reply src/api/java/com/minecolonies/api/crafting/RecipeStorage.java]: Yeah, cleaning this up

**Review by marchermans — DISMISSED** [review]

File: src/main/java/com/minecolonies/coremod/colony/crafting/CustomRecipe.java (hunk 23 above)
Comment (Mekle001) [reply src/main/java/com/minecolonies/coremod/colony/crafting/CustomRecipe.java]: Yeah. got it.

File: src/main/java/com/minecolonies/coremod/colony/requestsystem/requests/StandardRequests.java (hunk 24 above)
Comment (Mekle001) [reply src/main/java/com/minecolonies/coremod/colony/requestsystem/requests/StandardRequests.java]: Updated to be localizable, this now produces a string like "8 * Recipe: [Plain Oak Timber Frame]" for the crafting job entry in the request tree.

File: src/api/java/com/minecolonies/api/crafting/IRecipeStorage.java (hunk 10 above)
Comment (Mekle001) [reply src/api/java/com/minecolonies/api/crafting/IRecipeStorage.java]: Switched to Registry

File: src/main/java/com/minecolonies/coremod/client/gui/WindowListRecipes.java (hunk 20 above)
Comment (Mekle001) [reply src/main/java/com/minecolonies/coremod/client/gui/WindowListRecipes.java]: See switch to registry

File: src/api/java/com/minecolonies/api/crafting/AbstractRecipeType.java [hunk 25]
```diff
@@ -0,0 +1,41 @@
+package com.minecolonies.api.crafting;
+
+import java.util.List;
+import com.google.common.collect.ImmutableList;
+import net.minecraft.item.ItemStack;
+import net.minecraft.util.ResourceLocation;
+
+public abstract class AbstractRecipeType<R extends IRecipeStorage>
```
Comment (marchermans) [inline src/api/java/com/minecolonies/api/crafting/AbstractRecipeType.java:11]: Missing override indicators in this class.

**Review by marchermans — CHANGES_REQUESTED** [review]
Some nitpicking remains on one class.
Approved structurally and logically.

Rerequest a review once that class is marked properly.
After that this will probably be approved.

File: src/api/java/com/minecolonies/api/crafting/AbstractRecipeType.java (hunk 25 above)
Comment (someaddons) [reply src/api/java/com/minecolonies/api/crafting/AbstractRecipeType.java:11]: and class jdoc

File: src/api/java/com/minecolonies/api/crafting/MultiOutputRecipe.java [hunk 26]
```diff
@@ -0,0 +1,42 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+import com.google.common.collect.ImmutableList;
+
+import net.minecraft.item.ItemStack;
+import net.minecraft.util.ResourceLocation;
+
+public class MultiOutputRecipe extends AbstractRecipeType<IRecipeStorage>
```
Comment (someaddons) [inline src/api/java/com/minecolonies/api/crafting/MultiOutputRecipe.java:14]: class desc

File: src/api/java/com/minecolonies/api/crafting/AbstractRecipeType.java (hunk 25 above)
Comment (marchermans) [reply src/api/java/com/minecolonies/api/crafting/AbstractRecipeType.java:11]: Also that yes.

File: src/api/java/com/minecolonies/api/crafting/RecipeStorage.java [hunk 27]
```diff
@@ -228,13 +353,20 @@ public int hashCode()
...
-            final ItemStack container = stack.getItem().getContainerItem(stack);
-            if (!ItemStackUtils.isEmpty(container))
+            for (final ItemStack stack : input)
             {
-                container.setCount(stack.getCount());
-                secondaryStacks.add(container);
+                final ItemStack container = stack.getItem().getContainerItem(stack);
+                if (!ItemStackUtils.isEmpty(container))
+                {
+                    container.setCount(stack.getCount());
```
Comment (someaddons) [inline src/api/java/com/minecolonies/api/crafting/RecipeStorage.java:368]: if you're doing modifications in a simple checkForSpace() function you at least have to properly document/comment why that is there/whatfor it is. Or maybe even better rename the function

File: src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java [hunk 28]
```diff
@@ -238,7 +244,7 @@ protected int getWarehouseCount(ItemStorage item)
```
Comment (someaddons) [inline src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java:242]: while you're at it, can you change the above warehouse search to use the building managers saved warehouse list?

**Review by someaddons — COMMENTED** [review]
some small stuff

File: src/api/java/com/minecolonies/api/crafting/AbstractRecipeType.java (hunk 25 above)
Comment (Mekle001) [reply src/api/java/com/minecolonies/api/crafting/AbstractRecipeType.java:11]: Happy to add the jdoc, but there are no override indicators in this class because it doesn't extend or implement anything.

File: src/api/java/com/minecolonies/api/crafting/RecipeStorage.java (hunk 27 above)
Comment (Mekle001) [reply src/api/java/com/minecolonies/api/crafting/RecipeStorage.java:368]: Adding a few comments, renaming a variable, but not the function. It should be more clear now.

File: src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java (hunk 28 above)
Comment (Mekle001) [reply src/main/java/com/minecolonies/coremod/colony/buildings/AbstractBuildingWorker.java:242]: done

File: src/api/java/com/minecolonies/api/crafting/MultiOutputRecipe.java (hunk 26 above)
Comment (Mekle001) [reply src/api/java/com/minecolonies/api/crafting/MultiOutputRecipe.java:14]: Done

**Review by marchermans — APPROVED** [review]
